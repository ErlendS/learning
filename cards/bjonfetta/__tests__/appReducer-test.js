const test = require('ava');
const R = require('ramda')
const appReducer = require('../appReducer');
const { actions: playerActions } = require('../modules/player')
const { actions: tableActions } = require('../modules/tableStack');
const { actions: deckActions } = require('../modules/deck')
const { createPlayers, initPlayersHands } = require('../appActions');
const { generateDeck, generateShuffledDeck } = require('../../deck.js');
const utils = require('../bjonfettaUtils')



const _appReducer = R.flip(appReducer)

// test('appReducer - sortsPlayers', t => {
//   const initState = appReducer(undefined, undefined)
//   const deck = generateShuffledDeck() //unshuffled deck aka predictable
//
//   const state = R.compose(
//     _state => R.reduce(appReducer, _state, initPlayersHands(undefined, _state)),
//     _appReducer(deckActions.setDeck({ deck })),
//     _state => R.reduce(appReducer, _state, createPlayers({ n: 3 }, _state))
//   )(initState)
//
//   console.log(utils.sortPlayers(R.values(state.players)));
// })

test('appReducer - initPlayersHands connects to the deck and the players modules',t => {
  // generate players and deck.
  const initState = appReducer(undefined, undefined)
  const deck = generateDeck() //unshuffled deck aka predictable

  const state = R.compose(
    _state => R.reduce(appReducer, _state, initPlayersHands(undefined, _state)),
    _appReducer(deckActions.setDeck({ deck })),
    _state => R.reduce(appReducer, _state, createPlayers({ n: 3 }, _state))
  )(initState)
  console.log(R.values(state.players));
  t.snapshot(state)
  t.true(initState !== state)
})


test('appReducer - createPlayers connects to playersReducer', t => {
  const options = { n: 3 }
  const state = appReducer(undefined, undefined)
  const actual = R.reduce(appReducer, undefined , createPlayers(options, state))
  // const actual = appReducer(state, createPlayers(options, state))
  t.snapshot(actual)
})

test('appReducer - initializes state if inputs are undefined', t => {
  const actual = appReducer(undefined, undefined)
  t.snapshot(actual)

})
test('appReducer - connects to playerReducer', t => {
  const id1 = "Player_1"
  const id2 = "player_2"

  const actual = R.compose(
    _appReducer(playerActions.actionCreatePlayer({ id: id2})),
    _appReducer(playerActions.actionCreatePlayer({ id: id1}))
  )(undefined)

  t.snapshot(actual)
})


test('appReducer - connects to tableStackReducer', t => {
  const cards = ['H_2', 'S_13']
  const actual = appReducer(undefined, tableActions.actionAddCards({ cards }))
  t.snapshot(actual)
})


test('appReducer - connects to deckReducer', t => {
  const cards = ['S_3', 'D_3']
  const actual = appReducer(undefined, deckActions.addToDeck({ cards }))
  const expected = {
    tableStack: [],
    players: {},
    deck: ['S_3', 'D_3']
  }
  t.deepEqual(expected, actual)
  t.snapshot(actual)
})


test('appReducer - returns state if action is invalid', t => {
  const state = appReducer()
  const newState = appReducer(state, playerActions.actionCreatePlayer({ id: 'Player_1'}))
  const fakeAction = {
    payload: 'FAKE_NEWS',
    type: 'REMOVE_KARDS'
  }
  const actual = appReducer(newState, fakeAction)
  t.snapshot(actual)
})
