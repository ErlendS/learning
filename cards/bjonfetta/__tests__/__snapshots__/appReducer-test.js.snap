// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appActions - createPlayers connects to playersReducer 1`] = `
Object {
  "deck": Array [],
  "players": Object {
    "Player_0": Object {
      "cards": Object {
        "fd": Array [],
        "fu": Array [],
        "hand": Array [],
      },
      "id": "Player_0",
    },
    "Player_1": Object {
      "cards": Object {
        "fd": Array [],
        "fu": Array [],
        "hand": Array [],
      },
      "id": "Player_1",
    },
    "Player_2": Object {
      "cards": Object {
        "fd": Array [],
        "fu": Array [],
        "hand": Array [],
      },
      "id": "Player_2",
    },
  },
  "tableStack": Array [],
}
`;

exports[`appReducer - connects to deckReducer 1`] = `
Object {
  "deck": Array [
    "S_3",
    "D_3",
  ],
  "players": Object {},
  "tableStack": Array [],
}
`;

exports[`appReducer - connects to playerReducer 1`] = `
Object {
  "deck": Array [],
  "players": Object {
    "Player_1": Object {
      "cards": Object {
        "fd": Array [],
        "fu": Array [],
        "hand": Array [],
      },
      "id": "Player_1",
    },
    "player_2": Object {
      "cards": Object {
        "fd": Array [],
        "fu": Array [],
        "hand": Array [],
      },
      "id": "player_2",
    },
  },
  "tableStack": Array [],
}
`;

exports[`appReducer - connects to tableStackReducer 1`] = `
Object {
  "deck": Array [],
  "players": Object {},
  "tableStack": Array [
    "H_2",
    "S_13",
  ],
}
`;

exports[`appReducer - createPlayers connects to playersReducer 1`] = `
Object {
  "deck": Array [],
  "players": Object {
    "Player_0": Object {
      "cards": Object {
        "fd": Array [],
        "fu": Array [],
        "hand": Array [],
      },
      "id": "Player_0",
    },
    "Player_1": Object {
      "cards": Object {
        "fd": Array [],
        "fu": Array [],
        "hand": Array [],
      },
      "id": "Player_1",
    },
    "Player_2": Object {
      "cards": Object {
        "fd": Array [],
        "fu": Array [],
        "hand": Array [],
      },
      "id": "Player_2",
    },
  },
  "tableStack": Array [],
}
`;

exports[`appReducer - initPlayersHands connects to the deck and the players modules 1`] = `
Object {
  "deck": Array [
    "H_3",
    "H_4",
    "H_5",
    "H_6",
    "H_7",
    "H_8",
    "H_9",
    "H_10",
    "H_11",
    "H_12",
    "H_13",
    "H_14",
    "D_2",
    "D_3",
    "D_4",
    "D_5",
    "D_6",
    "D_7",
    "D_8",
    "D_9",
    "D_10",
    "D_11",
    "D_12",
    "D_13",
    "D_14",
  ],
  "players": Object {
    "Player_0": Object {
      "cards": Object {
        "fd": Array [
          "C_7",
          "C_8",
          "C_9",
        ],
        "fu": Array [
          "S_11",
          "S_12",
          "S_13",
        ],
        "hand": Array [
          "S_2",
          "S_3",
          "S_4",
        ],
      },
      "id": "Player_0",
    },
    "Player_1": Object {
      "cards": Object {
        "fd": Array [
          "C_10",
          "C_11",
          "C_12",
        ],
        "fu": Array [
          "S_14",
          "C_2",
          "C_3",
        ],
        "hand": Array [
          "S_5",
          "S_6",
          "S_7",
        ],
      },
      "id": "Player_1",
    },
    "Player_2": Object {
      "cards": Object {
        "fd": Array [
          "C_13",
          "C_14",
          "H_2",
        ],
        "fu": Array [
          "C_4",
          "C_5",
          "C_6",
        ],
        "hand": Array [
          "S_8",
          "S_9",
          "S_10",
        ],
      },
      "id": "Player_2",
    },
  },
  "tableStack": Array [],
}
`;

exports[`appReducer - initializes state if inputs are undefined 1`] = `
Object {
  "deck": Array [],
  "players": Object {},
  "tableStack": Array [],
}
`;

exports[`appReducer - returns state if action is invalid 1`] = `
Object {
  "deck": Array [],
  "players": Object {
    "Player_1": Object {
      "cards": Object {
        "fd": Array [],
        "fu": Array [],
        "hand": Array [],
      },
      "id": "Player_1",
    },
  },
  "tableStack": Array [],
}
`;
